### PR TITLE
[mlir] Update Enzyme MLIR pieces to LLVM 18

### PR DESCRIPTION
**Pull request by pengmai**
- Clean up warnings, clang-format
- Remove unused/outdated lower to llvm pass

This PR is an effort to chunk up #1549 into more manageable pieces.

### CODE REVIEW

File: enzyme/Enzyme/MLIR/enzymemlir-opt.cpp [hunk 1]
```diff
@@ -92,7 +94,7 @@ int main(int argc, char **argv) {
         PtrElementModel<LLVM::LLVMPointerType>>(*ctx);
     LLVM::LLVMArrayType::attachInterface<PtrElementModel<LLVM::LLVMArrayType>>(
         *ctx);
-    enzyme::CacheType::attachInterface<MemRefInsider>(*ctx);
+    // enzyme::CacheType::attachInterface<MemRefInsider>(*ctx);
```
Comment (ftynse) [inline enzyme/Enzyme/MLIR/enzymemlir-opt.cpp:97]: If we no longer need cache type inside memref, let's just delete this line.

File: enzyme/Enzyme/MLIR/Dialect/Ops.h [hunk 2]
```diff
@@ -16,7 +16,7 @@
 #include "mlir/Interfaces/ControlFlowInterfaces.h"
 #include "mlir/Interfaces/SideEffectInterfaces.h"
 
-#include "mlir/IR/DialectImplementation.h"
+#include "mlir/Bytecode/BytecodeOpInterface.h"
```
Comment (ftynse) [inline enzyme/Enzyme/MLIR/Dialect/Ops.h:19]: Are we using bytecode interface somewhere?

**Review by ftynse — APPROVED** [review]

File: enzyme/Enzyme/MLIR/Implementations/LLVMAutoDiffOpInterfaceImpl.cpp
Comment (wsmoses) [inline enzyme/Enzyme/MLIR/Implementations/LLVMAutoDiffOpInterfaceImpl.cpp:1]: This should probably be shadowType = gutils->getShadowType(…)

**Review by tgymnich — APPROVED** [review]

File: enzyme/Enzyme/MLIR/Implementations/SCFAutoDiffOpInterfaceImpl.cpp [hunk 3]
```diff
@@ -89,10 +88,10 @@ struct ForOpInterface
       if (!gutils->isConstantValue(std::get<0>(r)))
         nYields.push_back(gutils->invertPointerM(std::get<1>(r), builder));
     }
-    repFor.getBody()->push_back(
-        oldYield->create(oldYield->getLoc(), oldYield->getName(), TypeRange(),
-                         nYields, oldYield->getAttrs(),
-                         oldYield->getSuccessors(), oldYield->getNumRegions()));
+    repFor.getBody()->push_back(oldYield->create(
```
[the remark this answers is not in the document]
Comment (pengmai) [reply enzyme/Enzyme/MLIR/Implementations/SCFAutoDiffOpInterfaceImpl.cpp:91]: I think `clone` followed by `setOperands` is clearer in its intentions. I'm happy to change it to that for these parts.

File: enzyme/Enzyme/MLIR/Implementations/LLVMAutoDiffOpInterfaceImpl.cpp
Comment (pengmai) [reply enzyme/Enzyme/MLIR/Implementations/LLVMAutoDiffOpInterfaceImpl.cpp:1]: The MLIR gradient utils doesn't appear to have a `getShadowType` method, I'll need to either add one or do `shadowType = cast<AutoDiffTypeInterface>(...).getShadowType()`

File: enzyme/Enzyme/MLIR/Dialect/Ops.h (hunk 2 above)
Comment (pengmai) [reply enzyme/Enzyme/MLIR/Dialect/Ops.h:19]: My understanding (correct me if I'm wrong) is that this header is necessary to get the code to compile. It's in the [standalone dialect example](https://github.com/llvm/llvm-project/blob/main/mlir/examples/standalone/include/Standalone/StandaloneDialect.h) and it looks like the tablegen header files include references to `::mlir::DialectBytecodeWriter`, so compilation fails without it.

File: enzyme/Enzyme/MLIR/Implementations/LLVMAutoDiffOpInterfaceImpl.cpp
Comment (wsmoses) [reply enzyme/Enzyme/MLIR/Implementations/LLVMAutoDiffOpInterfaceImpl.cpp:1]: Ah fair, yeah we can do that later then.

**Review by wsmoses — APPROVED** [review]
Feel free to merge whenever you're ready!

File: enzyme/Enzyme/MLIR/Dialect/Ops.h (hunk 2 above)
Comment (ftynse) [reply enzyme/Enzyme/MLIR/Dialect/Ops.h:19]: Have you tried removing it and seeing what happens? (If it is indeed implied by the bare structure, we should make it much more obvious in the documentation).

Comment (pengmai) [reply enzyme/Enzyme/MLIR/Dialect/Ops.h:19]: Yup, removing it causes compilation to fail. Are you referring to the MLIR documentation or having a comment here?

Comment (ftynse) [reply enzyme/Enzyme/MLIR/Dialect/Ops.h:19]: MLIR documentation, but I don't necessarily expect _you_ to add that or make it a precondition for this to land. I'll investigate once this lands.

**Pull request merged**